src/python/grpcio/grpc/experimental/aio/_call.py

```diff
@@ -56,207 +81,331 @@ def __init__(self,
-            return False
-
-        code = grpc.StatusCode.CANCELLED
-        self._call_cancel_status.cancel(
-            _common.STATUS_CODE_TO_CYGRPC_STATUS_CODE[code],
-            details=Call._cancellation_details)
-        self._call.cancel()
-        self._details = Call._cancellation_details
-        self._code = code
-        self._state = _RpcState.CANCELLED
-        return True
+        raise NotImplementedError()
 
     def cancelled(self) -> bool:
-        """Returns if the RPC was cancelled.
-
-        Returns:
-          True if the requests was cancelled, False if not.
-        """
-        return self._state is _RpcState.CANCELLED
-
-    def running(self) -> bool:
-        """Returns if the RPC is running.
-
-        Returns:
-          True if the requests is running, False if it already terminated.
-        """
-        return not self.done()
+        return self._cancellation_future.done(
+        ) or self._code == grpc.StatusCode.CANCELLED
 
     def done(self) -> bool:
-        """Returns if the RPC has finished.
+        return self._status.done()
 
-        Returns:
-          True if the requests has finished, False is if still ongoing.
-        """
-        return self._state is not _RpcState.ONGOING
+    def add_callback(self, unused_callback) -> None:
+        pass
 
-    async def initial_metadata(self):
-        raise NotImplementedError()
+    def is_active(self) -> bool:
+        return self.done()
 
-    async def trailing_metadata(self):
-        raise NotImplementedError()
+    def time_remaining(self) -> float:
+        pass
 
-    async def code(self) -> grpc.StatusCode:
-        """Returns the `grpc.StatusCode` if the RPC is finished,
-        otherwise first waits until the RPC finishes.
+    async def initial_metadata(self) -> MetadataType:
+        return await self._initial_metadata
 
-        Returns:
-          The `grpc.StatusCode` status code.
-        """
-        if not self.done():
-            try:
-                await self
-            except (asyncio.CancelledError, AioRpcError):
-                pass
+    async def trailing_metadata(self) -> MetadataType:
+        return (await self._status).trailing_metadata()
 
+    async def code(self) -> grpc.StatusCode:
+        await self._status
         return self._code
 
     async def details(self) -> str:
-        """Returns the details if the RPC is finished, otherwise first waits till the
-        RPC finishes.
+        return (await self._status).details()
 
-        Returns:
-          The details.
+    async def debug_error_string(self) -> str:
+        return (await self._status).debug_error_string()
+
+    def _set_initial_metadata(self, metadata: MetadataType) -> None:
+        self._initial_metadata.set_result(metadata)
+
+    def _set_status(self, status: cygrpc.AioRpcStatus) -> None:
+        """Private method to set final status of the RPC.
+
+        This method may be called multiple time due to data race between local
```

Comment:
@lidizheng and I just had a conversation about the behavior here. I initially argued that if `cancel` is called on a completed RPC, the RPC's status should be overwritten with "cancelled". Conversely, if a couroutine is executed that would otherwise complete an RPC with a status from the server, but finds that the RPC has already been cancelled, it should leave the "cancelled" status alone. In other words, this would mean that a client application's request to cancel an RPC takes precedence over an RPC's actual completed status.

Lidi convinced me otherwise with the following argument. What does "cancelled" represent? Does it merely signify that the client application attempted to cancel? Or does it imply that a `RST_STREAM` frame has been sent to the server with a `CANCEL` error code? The former certainly seems less useful than the latter.